### ISSUE TITLE
Can correlations table contain attribute_id or 1_attribute_id that no longer exists?

### ISSUE THREAD

**Issue by github-germ** [MISP/MISP#6931]
2.4.133

Looking for guidance on this mystery. Many thanks!

We are seeing rows in the `correlations` db table that point to correlated attribute IDs that no longer exist in the `attributes` table.

**Is this a valid condition, and if not, any idea what might cause this situation?**

One far fetched hypothesis is that root cause may be our PyMISP experiments of hard deleting soft-deleted attributes, and perhaps the server is not removing the related row in the `correlations` table.

For example notice below `correlations.1_attribute_id` is not found in the `attributes` table, and that `correlations.value` does not match `value1` or `value2` for `correlations.attribute_id` in `attributes`. If our hypthesis is true, then I'd suspect our hard delete of a soft deleted attribute removed the row in the `attributes` table that matched this `correlations.value` but did not remove that row in the `correlations` table.

We welcome your wisdom... :-)

```
mysql> select * from correlations where id=47985836 \G
*************************** 1. row ***************************
                id: 47985836
             value: 146.88.232.72/1
        1_event_id: 180256
    1_attribute_id: 115449254
          event_id: 167315
      attribute_id: 70045868
            org_id: 29
      distribution: 4
    a_distribution: 5
  sharing_group_id: 13
a_sharing_group_id: 0
              date: 2020-02-03
              info: <REDACTED>
1 row in set (0.00 sec)

mysql> select * from attributes where id=115449254 \G
Empty set (0.00 sec)

mysql> select * from attributes where id=70045868 \G
*************************** 1. row ***************************
                 id: 70045868
           event_id: 167315
          object_id: 0
    object_relation: NULL
           category: Network activity
               type: url
             value1: 139.59.234.23
             value2:
             to_ids: 1
               uuid: a2ddc184-9e62-4961-8254-d09568a2f9f2
          timestamp: 1580954561
       distribution: 5
   sharing_group_id: 0
            comment:
            deleted: 0
disable_correlation: 0
         first_seen: NULL
          last_seen: NULL
1 row in set (0.00 sec)

mysql> 
```

**Comment by chrisinmtown** [comment]
This motivated me to check the table definition, and I was a little surprised to see that the MISP schema doesn't use a foreign key constraint, maybe for performance reasons?
```
CREATE TABLE IF NOT EXISTS `correlations` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `value` text COLLATE utf8_bin NOT NULL,
  `1_event_id` int(11) NOT NULL,
  `1_attribute_id` int(11) NOT NULL,
  `event_id` int(11) NOT NULL,
  `attribute_id` int(11) NOT NULL,
  `org_id` int(11) NOT NULL,
  `distribution` tinyint(4) NOT NULL,
  `a_distribution` tinyint(4) NOT NULL,
  `sharing_group_id` int(11) NOT NULL,
  `a_sharing_group_id` int(11) NOT NULL,
  `date` date NOT NULL,
  `info` text COLLATE utf8_bin NOT NULL,
  PRIMARY KEY (`id`),
  INDEX `event_id` (`event_id`),
  INDEX `1_event_id` (`1_event_id`),
  INDEX `attribute_id` (`attribute_id`),
  INDEX `1_attribute_id` (`1_attribute_id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin;
```

**Comment by iglocska** [comment]
Indeed, we rely purely on the linking done on the software level, though situations like these make me question that. My guess as to how this can happen in general:

The attribute save/delete pipelines first make the modifications to the attribute table and then initiate the correlation related actions. If the process would die in-between for whatever reason you would end up with orphaned correlations. 

I have checked if we have any such values in our operational instance and have 0 orphaned correlations, so not entirely sure how you ended up with them though, so I'm curious how to reproduce that.

**Comment by iglocska** [comment]
As for the hypothesis on the hard vs soft deletes, that shouldn't be the case. A soft delete already purges the correlations for the given attribute.

**Comment by chrisinmtown** [comment]
Looks like it's time for a cleanup script.  I don't see any PyMISP methods to add, update or delete correlations, am I reading api.py correctly?  If that's right, then I guess we have to resort to plain old SQL.

**Comment by github-germ** [comment]
@iglocska -- Thanks for the reply. It helps to know that indeed our `correlations` table is corrupted, and that it's not from hard deleting soft-deleted attributes. As to root cause, this will likely remain a mystery. We can write a SQL script that can monitor for this corrupted state in our instances, and if it recurs, see if we can isolate some root cause.

@chrisinmtown -- AFAIK, correlations are only managed on the server side transparent to PyMISP. The only "cleanup" we know for a corrupted `correlations` table is to disable/re-enable in the WebUI, and it rebuilds in the background over time.

**Comment by github-germ** [comment]
@iglocska -- would something like this SQL query suffice to discover if there's corruption? If you can recommend something better, that'd be great. Thanks.

```
mysql> select a.type,count(*) as cnt from correlations as c 
join attributes as a 
on c.attribute_id=a.id and UPPER(c.value) != UPPER(a.value1) 
and UPPER(c.value) != UPPER(a.value2) group by a.type;
+---------------+--------+
| type          | cnt    |
+---------------+--------+
| email-subject |      7 |
| ip-dst        | 576360 |
| ip-src        | 240858 |
| url           |      2 |
| ip-dst|port   |    176 |
| ip-src|port   |      4 |
| mutex         |      1 |
+---------------+--------+
7 rows in set (2 min 0.24 sec)

mysql>
```

**Comment by packet-rat** [comment]
re_: "The attribute save/delete pipelines first make the modifications to the attribute table and then initiate the correlation related actions. If the process would die in-between for whatever reason you would end up with orphaned correlations.

I have checked if we have any such values in our operational instance and have 0 orphaned correlations, so not entirely sure how you ended up with them though, so I'm curious how to reproduce that."_

I've been running pymisp based ingestion tests which create 100s of Events and 10,000s of Attributes.   I have experienced the dreaded "black-hole" and other obtuse error messages when deleting Events via the WebUI (i.e., select a screen's worth of Events and click the Delete Icon, wash...rinse...repeat).  I wish I could remember when I've seen these WebUI Errors over the past months - I can't.

I can state that the xxx Event (the patient zero that start us down the Correlations investigation) is indeed one that I've created and deleted many times.

In my proto-typing, I've also been creating/deleting Events in one of three modes:

(1) A single Event for a given xxx Enclave that just get's the list of Attributes from all the xxx Reports and concatenates them to that Event.

(2) A single Event for a given xxx Enclave that adds a new MISP xxx Report custom Object for each xxx Report in that Enclave.

(3) A new MISP Event for every xxx Report in a given Enclave.

While iterating through ingestion of ~24 xxx Enclaves (each with their own sources/formats) we've created multiple instantiations of these to see which methods convey the most context/tagging - balanced against impacts to Correlations in general.  At times there have Events that contain both flat Attributes and xxx Report MISP Objects concurrent for the same data.

So two key "take away"s are that:

(1) I frequently see errors in the MISP UI when trying to ingest/import/delete data.  Almost all error messages are obtuse at best.  In most cases a refresh or return to a page indicate that despite the error, the action seemed to complete.   This would definitely be the case when I've gotten errors while deleting single/multiple Events from the WebUI.

(2) There's been a lot of churn from the extensive testing I've been doing developing the custom xxx Enclave ingestion processes.    I've created/deleted 10s of millions of Events/Attributes over the past ~10 months.  A great deal of these have been deleted using our internal Bulk Deletion utility.

If I had to guess (1) is most likely the failure mechanism as it really aligns with Andras's description of what can lead to this orphaned Correlations condition.

**Comment by chrisinmtown** [comment]
@iglocska I sure hope you will extend the MISP schema with foreign-key constraints from the correlations table to the attributes table.  Then if your code does the wrong thing you'll instantly know about it because the database will not allow the operation. Even at some cost in performance I think having the database defend itself is far preferable to the situation we have now, where an unknown behavior of the code leaves bad data for people to dig out by hand, tediously, after it becomes an enormous problem.

**Comment by github-germ** [comment]
@iglocska -- With 2.4.143 seeing what to me appears to be correlations table corruption.
1. Does My query seem correct?
2. If indeed there is corruption again, how can we look for root cause?
```
mysql> select a.type,count(*) as cnt
    -> from correlations as c
    -> join attributes as a
    ->  on c.attribute_id=a.id and upper(c.value) != upper(a.value1)
    ->     and upper(c.value) != upper(a.value2)
    -> group by a.type;
+------------------------+-------+
| type                   | cnt   |
+------------------------+-------+
| url                    |  6122 |
| ip-dst                 | 13559 |
| domain                 |  4742 |
| hostname               |   140 |
| md5                    |   220 |
| link                   |    36 |
| email-src              |    59 |
| email-subject          |    22 |
| email-src-display-name |    50 |
| email-header           |    42 |
| email-reply-to         |    24 |
| sha256                 |   191 |
| ip-src                 |    27 |
| threat-actor           |     1 |
| filename               |    10 |
| comment                |    14 |
| email-attachment       |     6 |
| campaign-id            |     4 |
| ssdeep                 |   168 |
| sha1                   |   195 |
+------------------------+-------+
20 rows in set (19.82 sec)

mysql>
```

**Comment by RichieB2B** [comment]
@github-germ You can add foreign keys constraints to the correlations table as @chrisinmtown suggests. Then look for errors in the MISP logs to find the root cause.
